src/messenger_packets.md

```diff
@@ -0,0 +1,147 @@
+
+When messenger layer is called by user using FILE_CONTROL, it checks:
+
+- Sending direction value is `0` or `1` means 0 = sender, 1 = receiver, others = error
+- Retrieves `file transfer object` which holds info of file transferring, if it fails then send FileControl packet to a friend to kill the session.
+- `file transfer object` is managed by `file_id` and `friend_id` and `sending direction`.
+
+FILE_CONTROL do following things:
+
+- Accept: Accept the request of file sending from a friend. It does:
+    - checks if status is not `Accepted` if it is not `Accepted` then
+        - change status to `Transferring`
+    - else checks:
+        - if status is `Pause by friend` then toggle status of `Pause by friend`
+        - else error because friend asked me to resume file transfer that wasn't paused.
+    - updates `file transfer object`
+
+- Pause: Pause the transfer. It does:
+    - checks if pause status is `Paused by friend` or transfer status is not `Transferring` then
+    error because friend asked me to pause transfer that is already paused.
+    - toggles pause status
+    - updates `file transfer object`
+    
+- Kill: Kill the transfer session. It does:
+    - updates `file transfer object`
+    - set transfer status to `None`
+    - decreases count of friend's number of sending files
+    
+- Seek: Seek to the position. It does:
+    - checks if the position is u64, means position is 8 bytes long
+        - else error because expected 8 bytes
+    - checks if the transfer status is `Accepted` and the sender of this packet is receiver of file data
+        - else error because Seek must be used to session `Accepted` and Seek packet can only be sent by receiver to seek before resuming broken transfers.
+    - checks if the position exceeds the file size, if it exceeds
+        - error because seek position exceeds file size.
+    - set `requested position` and `transferred position` to seek position.
+- else error because invalid file control command.
+
+## TYPING
+
+This packet is used to transmit sender's typing status to a friend.
+
+Serialized form:
+
+Length    | Content
+--------- | ------
+`1`       | `0x33`
+`1`       | Typing status(0 = not typing, 1 = typing)
+
+It checks if the data is one bytes, and update typing status to the value of data.
+
+## ACTION
+
+This packet is used to transmit sender's action message to a friend.
+Here, action message is a something like an IRC action.
+
+Serialized form:
+
+Length    | Content
+--------- | ------
+`1`       | `0x41`
+`0..1372` | UTF8 byte string
+
+It checks if the action message is null, if null then do nothing else sends the action message to the friend.
+
+## MESSAGE
+
+This packet is used to transmit sender's message to a friend.
+
+Serialized form:
+
+Length    | Content
+--------- | ------
+`1`       | `0x40`
+`0..1372` | UTF8 byte string
+
+It checks if the message is null, if null then do nothing else sends the message to the friend.
```

Comment:
> It checks if the message is null, if null then do nothing else sends the message to the friend.

Better change to: `there is no zero-length or empty string allowed`.

BTW is it a C string or Pascal string?

Reply:
It is C string